File: Sources/LocalScreenTranslator/TranslationService.swift
import Foundation

enum TranslationDirection {
    case englishToChinese
    case chineseToEnglish

    var sourceLanguageName: String {
        switch self {
        case .englishToChinese:
            "English"
        case .chineseToEnglish:
            "Chinese"
        }
    }

    var targetLanguageName: String {
        switch self {
        case .englishToChinese:
            "Simplified Chinese"
        case .chineseToEnglish:
            "English"
        }
    }
}

@MainActor
final class TranslationRouter {
    private let settings: AppSettings
    private let englishToChineseApple: TranslationCoordinator
    private let chineseToEnglishApple: TranslationCoordinator
    private let localAIService = OMLXTranslationService()

    init(
        settings: AppSettings,
        englishToChineseApple: TranslationCoordinator,
        chineseToEnglishApple: TranslationCoordinator
    ) {
        self.settings = settings
        self.englishToChineseApple = englishToChineseApple
        self.chineseToEnglishApple = chineseToEnglishApple
    }

    func translate(_ texts: [String], direction: TranslationDirection) async throws -> [String] {
        switch settings.translationEngine {
        case .appleTranslation:
            return try await translateWithApple(texts, direction: direction)
        case .localAIWithAppleFallback:
            do {
                return try await localAIService.translate(texts, direction: direction, settings: settings)
            } catch {
                guard settings.appleTranslationFallbackEnabled else {
                    throw TranslationFailure(message: "本地 AI 翻译失败：\(error.localizedDescription)")
                }

                do {
                    return try await translateWithApple(texts, direction: direction)
                } catch {
                    throw TranslationFailure(message: "本地 AI 翻译失败，备用 Apple 机翻也失败：\(error.localizedDescription)")
                }
            }
        }
    }

    private func translateWithApple(_ texts: [String], direction: TranslationDirection) async throws -> [String] {
        let coordinator = direction == .chineseToEnglish ? chineseToEnglishApple : englishToChineseApple
        return try await coordinator.translate(texts)
    }
}

final class OMLXTranslationService {
    func translate(
        _ texts: [String],
        direction: TranslationDirection,
        settings: AppSettings
    ) async throws -> [String] {
        let cleanTexts = texts.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
        guard cleanTexts.contains(where: { !$0.isEmpty }) else { return texts }

        do {
            return try await translateBatch(cleanTexts, direction: direction, settings: settings)
        } catch {
            var results: [String] = []
            results.reserveCapacity(cleanTexts.count)
            for text in cleanTexts {
                guard !text.isEmpty else {
                    results.append("")
                    continue
                }
                let translated = try await translateSingle(text, direction: direction, settings: settings)
                results.append(translated)
            }
            return results
        }
    }

    private func translateBatch(
        _ texts: [String],
        direction: TranslationDirection,
        settings: AppSettings
    ) async throws -> [String] {
        let numberedText = texts.enumerated()
            .map { index, text in "\(index + 1). \(text)" }
            .joined(separator: "\n")
        let prompt = """
        Translate each numbered item from \(direction.sourceLanguageName) to \(direction.targetLanguageName).
        Return only the translations, one per line, with the same numbering. Do not add explanations.

        \(numberedText)
        """
        let response = try await request(prompt: prompt, settings: settings)
        return try parseNumbered(response, expectedCount: texts.count)
    }

    private func translateSingle(
        _ text: String,
        direction: TranslationDirection,
        settings: AppSettings
    ) async throws -> String {
        let prompt = """
        Translate the following text from \(direction.sourceLanguageName) to \(direction.targetLanguageName).
        Return only the translated text. Do not add explanations.

        \(text)
        """
        return try await request(prompt: prompt, settings: settings)
    }

    private func request(prompt: String, settings: AppSettings) async throws -> String {
        let endpoint = try chatCompletionsURL(from: settings.localAIBaseURL)
        let model = settings.localAIModel
        guard !model.isEmpty else {
            throw TranslationFailure(message: "本地 AI 模型名为空。")
        }

        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if !settings.localAIAPIKey.isEmpty {
            request.setValue("Bearer \(settings.localAIAPIKey)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = try JSONEncoder().encode(ChatCompletionRequest(
            model: model,
            messages: [
                ChatMessage(role: "system", content: "You are a translation engine. Output translated text only."),
                ChatMessage(role: "user", content: prompt)
            ],
            temperature: 0
        ))

        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = settings.localAITimeout
        configuration.timeoutIntervalForResource = settings.localAITimeout
        let session = URLSession(configuration: configuration)
        let (data, response) = try await session.data(for: request)

        if let httpResponse = response as? HTTPURLResponse,
           !(200...299).contains(httpResponse.statusCode) {
            let message = String(data: data, encoding: .utf8) ?? HTTPURLResponse.localizedString(forStatusCode: httpResponse.statusCode)
            throw TranslationFailure(message: "本地 AI 返回 HTTP \(httpResponse.statusCode)：\(message)")
        }

        let decoded = try JSONDecoder().decode(ChatCompletionResponse.self, from: data)
        guard let content = decoded.choices.first?.message.content.trimmingCharacters(in: .whitespacesAndNewlines),
              !content.isEmpty else {
            throw TranslationFailure(message: "本地 AI 没有返回译文。")
        }
        return content
    }

    private func chatCompletionsURL(from baseURL: String) throws -> URL {
        let trimmed = baseURL.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            throw TranslationFailure(message: "本地 AI Base URL 为空。")
        }

        let normalized = trimmed.hasSuffix("/") ? String(trimmed.dropLast()) : trimmed
        guard let url = URL(string: normalized + "/chat/completions") else {
            throw TranslationFailure(message: "本地 AI Base URL 无效。")
        }
        return url
    }

    private func parseNumbered(_ response: String, expectedCount: Int) throws -> [String] {
        let lines = response
            .components(separatedBy: .newlines)
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }

        var results = Array(repeating: "", count: expectedCount)
        var matchedCount = 0
        let pattern = #"^\s*(\d+)[\.\)、\)]\s*(.+)$"#
        let regex = try NSRegularExpression(pattern: pattern)

        for line in lines {
            let nsRange = NSRange(line.startIndex..<line.endIndex, in: line)
            guard let match = regex.firstMatch(in: line, range: nsRange),
                  match.numberOfRanges == 3,
                  let indexRange = Range(match.range(at: 1), in: line),
                  let textRange = Range(match.range(at: 2), in: line),
                  let index = Int(line[indexRange]),
                  (1...expectedCount).contains(index)
            else {
                continue
            }

            results[index - 1] = String(line[textRange]).trimmingCharacters(in: .whitespacesAndNewlines)
            matchedCount += 1
        }

        guard matchedCount == expectedCount,
              !results.contains(where: { $0.isEmpty }) else {
            throw TranslationFailure(message: "本地 AI 批量译文格式无法解析。")
        }

        return results
    }
}

private struct ChatCompletionRequest: Encodable {
    let model: String
    let messages: [ChatMessage]
    let temperature: Double
}

private struct ChatMessage: Codable {
    let role: String
    let content: String
}

private struct ChatCompletionResponse: Decodable {
    let choices: [ChatChoice]
}

private struct ChatChoice: Decodable {
    let message: ChatMessage
}

File: windows/Huyi.Windows/Services/TranslationService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;
using Huyi.Windows.Models;

namespace Huyi.Windows.Services;

public sealed class TranslationService
{
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web);

    public async Task<IReadOnlyList<string>> TranslateAsync(
        IReadOnlyList<string> texts,
        TranslationDirection direction,
        AppSettings settings,
        CancellationToken cancellationToken = default)
    {
        var cleanTexts = texts.Select(text => text.Trim()).ToArray();
        if (cleanTexts.All(string.IsNullOrWhiteSpace))
        {
            return texts.ToArray();
        }

        try
        {
            return await TranslateBatchAsync(cleanTexts, direction, settings, cancellationToken);
        }
        catch
        {
            var results = new List<string>(cleanTexts.Length);
            foreach (var text in cleanTexts)
            {
                results.Add(string.IsNullOrWhiteSpace(text)
                    ? ""
                    : await TranslateSingleAsync(text, direction, settings, cancellationToken));
            }
            return results;
        }
    }

    private async Task<IReadOnlyList<string>> TranslateBatchAsync(
        IReadOnlyList<string> texts,
        TranslationDirection direction,
        AppSettings settings,
        CancellationToken cancellationToken)
    {
        var numbered = string.Join(Environment.NewLine, texts.Select((text, index) => $"{index + 1}. {text}"));
        var prompt = $"""
            Translate each numbered item from {SourceName(direction)} to {TargetName(direction)}.
            Return only the translations, one per line, with the same numbering. Do not add explanations.

            {numbered}
            """;

        var response = await RequestAsync(prompt, settings, cancellationToken);
        return ParseNumbered(response, texts.Count);
    }

    private async Task<string> TranslateSingleAsync(
        string text,
        TranslationDirection direction,
        AppSettings settings,
        CancellationToken cancellationToken)
    {
        var prompt = $"""
            Translate the following text from {SourceName(direction)} to {TargetName(direction)}.
            Return only the translated text. Do not add explanations.

            {text}
            """;

        return await RequestAsync(prompt, settings, cancellationToken);
    }

    private async Task<string> RequestAsync(
        string prompt,
        AppSettings settings,
        CancellationToken cancellationToken)
    {
        if (string.IsNullOrWhiteSpace(settings.LmStudioModel))
        {
            throw new InvalidOperationException("本地 AI 模型名为空。");
        }

        using var client = new HttpClient
        {
            Timeout = TimeSpan.FromSeconds(Math.Max(1, settings.LmStudioTimeoutSeconds))
        };

        using var request = new HttpRequestMessage(HttpMethod.Post, ChatCompletionsUrl(settings.LmStudioBaseUrl));
        request.Content = new StringContent(JsonSerializer.Serialize(new ChatCompletionRequest(
            settings.LmStudioModel,
            [
                new ChatMessage("system", "You are a translation engine. Output translated text only."),
                new ChatMessage("user", prompt)
            ],
            0
        ), JsonOptions), Encoding.UTF8, "application/json");

        if (!string.IsNullOrWhiteSpace(settings.LmStudioApiKey))
        {
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", settings.LmStudioApiKey);
        }

        using var response = await client.SendAsync(request, cancellationToken);
        var body = await response.Content.ReadAsStringAsync(cancellationToken);
        if (!response.IsSuccessStatusCode)
        {
            throw new InvalidOperationException($"本地 AI 返回 HTTP {(int)response.StatusCode}：{body}");
        }

        var decoded = JsonSerializer.Deserialize<ChatCompletionResponse>(body, JsonOptions);
        var content = decoded?.Choices.FirstOrDefault()?.Message.Content.Trim();
        if (string.IsNullOrWhiteSpace(content))
        {
            throw new InvalidOperationException("本地 AI 没有返回译文。");
        }
        return content;
    }

    private static Uri ChatCompletionsUrl(string baseUrl)
    {
        var trimmed = baseUrl.Trim().TrimEnd('/');
        if (string.IsNullOrWhiteSpace(trimmed))
        {
            throw new InvalidOperationException("本地 AI Base URL 为空。");
        }

        return Uri.TryCreate($"{trimmed}/chat/completions", UriKind.Absolute, out var uri)
            ? uri
            : throw new InvalidOperationException("本地 AI Base URL 无效。");
    }

    private static IReadOnlyList<string> ParseNumbered(string response, int expectedCount)
    {
        var results = Enumerable.Repeat("", expectedCount).ToArray();
        var matched = 0;
        var regex = new Regex(@"^\s*(\d+)[\.\)、\)]\s*(.+)$", RegexOptions.Compiled);

        foreach (var line in response.Split('\n').Select(line => line.Trim()).Where(line => line.Length > 0))
        {
            var match = regex.Match(line);
            if (!match.Success || !int.TryParse(match.Groups[1].Value, out var index))
            {
                continue;
            }

            if (index < 1 || index > expectedCount)
            {
                continue;
            }

            results[index - 1] = match.Groups[2].Value.Trim();
            matched++;
        }

        if (matched != expectedCount || results.Any(string.IsNullOrWhiteSpace))
        {
            throw new InvalidOperationException("本地 AI 批量译文格式无法解析。");
        }

        return results;
    }

    private static string SourceName(TranslationDirection direction) =>
        direction == TranslationDirection.EnglishToChinese ? "English" : "Chinese";

    private static string TargetName(TranslationDirection direction) =>
        direction == TranslationDirection.EnglishToChinese ? "Simplified Chinese" : "English";

    private sealed record ChatCompletionRequest(string Model, IReadOnlyList<ChatMessage> Messages, double Temperature);
    private sealed record ChatMessage(string Role, string Content);
    private sealed record ChatCompletionResponse(IReadOnlyList<ChatChoice> Choices);
    private sealed record ChatChoice(ChatMessage Message);
}
